src/Core/Algorithms/Math/ComputeTensorUncertaintyAlgorithm.cc

```diff
@@ -0,0 +1,443 @@
+  void computeMeanTensors();
+  Dyadic3DTensor computeMeanMatrixAverage(int index) const;
+  Dyadic3DTensor computeMeanLinearInvariant(int t) const;
+  Dyadic3DTensor computeMeanLogEuclidean(int t) const;
+};
+
+void ComputeTensorUncertaintyAlgorithmImpl::setInvariantMethod(AlgoOption method)
+{
+  invariantMethod_ = method;
+}
+
+void ComputeTensorUncertaintyAlgorithmImpl::setOrientationMethod(AlgoOption method)
+{
+  orientationMethod_ = method;
+}
+
+FieldHandle ComputeTensorUncertaintyAlgorithmImpl::getMeanTensors() const
+{
+  FieldInformation ofinfo("PointCloudMesh", 0, "Tensor");
+  auto ofield = CreateField(ofinfo);
+  auto mesh = ofield->vmesh();
+  auto field = ofield->vfield();
+
+  std::vector<VMesh::index_type> meshIndices(fieldSize_);
+  for (size_t i = 0; i < fieldSize_; ++i)
+    meshIndices[i] = mesh->add_point(points_[0][i]);
+
+  field->resize_fdata();
+
+  for (size_t i = 0; i < fieldSize_; ++i)
+    field->set_value(eigenTensorToScirunTensor(meanTensors_[i]), meshIndices[i]);
+
+  return ofield;
+}
+
+void ComputeTensorUncertaintyAlgorithmImpl::computeMeanTensors()
+{
+  meanTensors_ = std::vector<Dyadic3DTensor>(fieldSize_);
+  if (invariantMethod_.option_ == orientationMethod_.option_)
+  {
+    if (orientationMethod_.option_ == "Matrix Average") {
+      for (size_t t = 0; t < fieldSize_; ++t)
+        meanTensors_[t] = computeMeanMatrixAverage(t);
+    }
+    else if (orientationMethod_.option_ == "Log-Euclidean") {
+      for (size_t t = 0; t < fieldSize_; ++t)
+        meanTensors_[t] = computeMeanLogEuclidean(t);
+    }
+  }
+  else
+  {
+    std::vector<Dyadic3DTensor> orientationTensors(fieldSize_);
+    std::vector<Dyadic3DTensor> invariantTensors(fieldSize_);
+    if (orientationMethod_.option_ == "Matrix Average") {
+      for (size_t t = 0; t < fieldSize_; ++t)
+        orientationTensors[t] = computeMeanMatrixAverage(t);
+    }
+    else if (orientationMethod_.option_ == "Log-Euclidean") {
+      for (size_t t = 0; t < fieldSize_; ++t)
+        orientationTensors[t] = computeMeanLogEuclidean(t);
+    }
+
+    if (invariantMethod_.option_ == "Matrix Average"){
+      for (size_t t = 0; t < fieldSize_; ++t)
+        invariantTensors[t] = computeMeanMatrixAverage(t);
+    }
+    else if (invariantMethod_.option_ == "Log-Euclidean") {
+      for (size_t t = 0; t < fieldSize_; ++t)
+        invariantTensors[t] = computeMeanLogEuclidean(t);
+    }
+    else if (invariantMethod_.option_ == "Linear Invariant") {
+      for (size_t t = 0; t < fieldSize_; ++t)
+      {
+        invariantTensors[t] = computeMeanLinearInvariant(t);
+      }
+    }
+
+    for (size_t t = 0; t < fieldSize_; ++t)
+    {
+      invariantTensors[t].setDescendingRHSOrder();
+      orientationTensors[t].setDescendingRHSOrder();
+      meanTensors_[t] = Dyadic3DTensor(orientationTensors[t].getEigenvectors(),
+                                       invariantTensors[t].getEigenvalues());
+      meanTensors_[t].setDescendingRHSOrder();
+    }
+  }
+}
```

Comment:
At least break this up into two methods, one for the `if (invariantMethod_.option_ == orientationMethod_.option_)` case and one for the else.

Reply:
sounds good